org.eclipse.xtext.ecore/src/org/eclipse/xtext/ecore/EcoreQualifiedNameProvider.java

```diff
@@ -49,7 +40,28 @@ public QualifiedName getFullyQualifiedName(final EObject obj) {
 			@Override
 			public QualifiedName get() {
 				EObject temp = obj;
-				String name = nameDispatcher.invoke(temp);
+				String name = new EcoreSwitch<String>() {
```

Comment:
This might even be slower than before since it'll call from the most special type up to the default case for all objects that are not in here. Did you measure this e.g. with JMH?

Since it is an anonymous inner class, I don't see how this could be of real benefit. It is not really reusable.

If it is provably faster, I'd rather make EcoreQualifiedNameProvider an `@Singleton` and inherit from the EcoreSwitch with more elaborate overrides for the objects that are often present in an ePackage but don't have a name.

Reply:
> This might even be slower than before since it'll call from the most special type up to the default case for all objects that are not in here. Did you measure this e.g. with JMH?
> 

I never used JMH so I wouldn't know where to start from (but I'd like to learn of course, if you have some suggestions). Concerning the first part, do you mean that it's best to have cases for EClass, EDataType instead of EClassifier?

> Since it is an anonymous inner class, I don't see how this could be of real benefit. It is not really reusable.

Something like the private field as the polymorphic dispatcher used to be?

> 
> If it is provably faster, I'd rather make EcoreQualifiedNameProvider an `@Singleton` and inherit from the EcoreSwitch with more elaborate overrides for the objects that are often present in an ePackage but don't have a name.

This last part is not clear to me :)

Reply:
Being unclear is hobby of mine I'm afraid :)

I was thinking
```java
@Singleton
public class EcoreQualifiedNameProvider extends EcoreSwitch<QualifiedName> implements IQualifiedNameProvider {
  // no caching no nothing, just fairly optimized computation of the QualifiedName for few types that are relevant
  // e.g.
  @Override
  public QualifiedName caseEStructuralFeature(EStructuralFeature object) {
    // might even go fancier here with a QualifiedName.builder(..)
    return doSwitch(object.getEContainingClass()).append(object.getName());
  }
  ..
}
```